File: sonar-go-checks/src/main/java/org/sonar/go/checks/IdenticalBinaryOperandCheck.java
/*
 * SonarSource Go
 * Copyright (C) 2018-2025 SonarSource SA
 * mailto:info AT sonarsource DOT com
 *
 * This program is free software; you can redistribute it and/or
 * modify it under the terms of the Sonar Source-Available License Version 1, as published by SonarSource SA.
 *
 * This program is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.
 * See the Sonar Source-Available License for more details.
 *
 * You should have received a copy of the Sonar Source-Available License
 * along with this program; if not, see https://sonarsource.com/license/ssal/
 */
package org.sonar.go.checks;

import java.util.EnumSet;
import java.util.Set;
import org.sonar.check.Rule;
import org.sonar.plugins.go.api.BinaryExpressionTree;
import org.sonar.plugins.go.api.checks.GoCheck;
import org.sonar.plugins.go.api.checks.InitContext;
import org.sonar.plugins.go.api.checks.SecondaryLocation;

import static org.sonar.go.utils.ExpressionUtils.containsPlaceHolder;
import static org.sonar.go.utils.ExpressionUtils.skipParentheses;
import static org.sonar.go.utils.SyntacticEquivalence.areEquivalent;
import static org.sonar.plugins.go.api.BinaryExpressionTree.*;

@Rule(key = "S1764")
public class IdenticalBinaryOperandCheck implements GoCheck {

  public static final String MESSAGE = "Correct one of the identical sub-expressions on both sides of this operator.";
  private static final Set<Operator> EXCEPTIONS = EnumSet.of(Operator.PLUS, Operator.TIMES, Operator.BITWISE_SHL);

  @Override
  public void initialize(InitContext init) {
    init.register(BinaryExpressionTree.class, (ctx, tree) -> {
      if (!EXCEPTIONS.contains(tree.operator())
        && !containsPlaceHolder(tree)
        && areEquivalent(skipParentheses(tree.leftOperand()), skipParentheses(tree.rightOperand()))) {
        ctx.reportIssue(tree.rightOperand(), MESSAGE, new SecondaryLocation(tree.leftOperand()));
      }
    });
  }

}

File: sonar-go-checks/src/test/resources/checks/IdenticalBinaryOperandCheck/IdenticalBinaryOperandCheck.go
package IdenticalBinaryOperandCheck

func main() {
  x == 1
   1 == 1 // Noncompliant {{Correct one of the identical sub-expressions on both sides of this operator.}}
// ^>   ^
  1 == (1)           // Noncompliant {{Correct one of the identical sub-expressions on both sides of this operator.}}
  (1 + 2) == 1+2     // Noncompliant
  (1 + 2) == (1 + 2) // Noncompliant
  (1 + 2) == (1 + 2 + 3)
  (x ==
// ^>
    x) // Noncompliant
 // ^
  1 == 2
  x <= x   // Noncompliant
  _x <= _x // Noncompliant
  x_ <= x
  _x <= x
  x <= _
  _ <= y
  x >> x // Noncompliant
  // Exceptions:
  1 << 1
  x << x
  x + x
  x * x
  // The following one is not a binary expression, but an assignment expression.
  x = x
}
